query/src/pipelines/new/processors/transforms/hash_join/hash_table.rs

```diff
@@ -0,0 +1,321 @@
+use crate::pipelines::new::processors::transforms::hash_join::row::RowSpace;
+use crate::sessions::QueryContext;
+
+/// Concurrent hash table for hash join.
+pub trait HashJoinState: Send + Sync {
+    /// Build hash table with input DataBlock
+    fn build(&self, input: DataBlock) -> Result<()>;
+
+    /// Probe the hash table and retrieve matched rows as DataBlocks
+    fn probe(&self, input: &DataBlock) -> Result<Vec<DataBlock>>;
+
+    /// Attach to state
+    fn attach(&self) -> Result<()>;
+
+    /// Detach to state
+    fn detach(&self) -> Result<()>;
+
+    /// Is building finished.
+    fn is_finished(&self) -> Result<bool>;
+
+    /// Finish building hash table, will be called only once as soon as all handles
+    /// have been detached from current state.
+    fn finish(&self) -> Result<()>;
+}
+
+pub struct ChainHashTable {
+    /// Reference count
+    ref_count: Mutex<usize>,
+    is_finished: Mutex<bool>,
+
+    build_expressions: Vec<Expression>,
+    probe_expressions: Vec<Expression>,
+
+    ctx: Arc<QueryContext>,
+
+    hash_method_kind: HashMethodKind,
+    /// A shared big hash table stores all the rows from build side
+    hash_table: RwLock<Vec<Option<RowPtr>>>,
+    row_space: RowSpace,
+}
+
+impl ChainHashTable {
+    pub fn try_create(
+        build_expressions: Vec<Expression>,
+        probe_expressions: Vec<Expression>,
+        build_data_schema: DataSchemaRef,
+        _probe_data_schema: DataSchemaRef,
+        ctx: Arc<QueryContext>,
+    ) -> Result<Self> {
+        // TODO: support complicated expressions
+        let build_keys = build_expressions
+            .iter()
+            .map(|expr| expr.to_data_type(&build_data_schema))
+            .collect::<Result<Vec<DataTypeImpl>>>()?;
+        let hash_method_kind = DataBlock::choose_hash_method_with_types(&build_keys)?;
+        Ok(Self {
+            row_space: RowSpace::new(build_data_schema),
+            ref_count: Mutex::new(0),
+            is_finished: Mutex::new(false),
+            build_expressions,
+            probe_expressions,
+            ctx,
+            hash_method_kind,
+            hash_table: RwLock::new(vec![]),
+        })
+    }
+
+    fn get_matched_ptrs(&self, hash_key: u64) -> Vec<RowPtr> {
+        let hash_table = self.hash_table.read().unwrap();
+        let mut ptr: Option<RowPtr> = hash_table[hash_key as usize];
+        let mut result: Vec<RowPtr> = vec![];
+
+        while let Some(v) = ptr {
+            result.push(v);
+            ptr = self.row_space.get_next(&v);
+        }
+
+        result
+    }
+
+    fn hash(&self, columns: &[ColumnRef], row_count: usize) -> Result<HashVector> {
+        let columns_vec: Vec<ColumnRef> = columns.to_vec();
+        let columns: Vec<&ColumnRef> = columns_vec.iter().collect();
+        Ok(match &self.hash_method_kind {
+            HashMethodKind::Serializer(method) => method
+                .build_keys(&columns, row_count)?
```

Comment:
I did not find any reason to use `hash_method_kind`.
Evaluate [`xxhash64`](https://github.com/datafuselabs/databend/blob/752ecb112c68d22d65298dd309aab41a9ceaa7bd/common/functions/src/scalars/hashes/hash.rs) may be more simple way to do it.

Reply:
I was trying to reuse the `HashMethod`, but it seems the need of hashing is different from `Aggregate`.

It only needs a hashed `u64` value here. Maybe I can just compute hash for each column and combine them together?

Reply:
`HashMethod ` is optimized in the case of group aggregation, it produces unique keys and leads to more memory.

Now you only need a hashed u64 value, so `HashMethod` may be useless unless you store your state using HashTable.

Reply:
I got your point, now hash function may not work for multi-columns.  Let's keep it as now and refactor it in the future.

Reply:
@sundy-li I have remove the `HashMethod` and use hash `Function` instead. PTAL